### PR TITLE
GoMod: Fix analyzing projects with unused dependencies

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits.

**Note: Must be merged after** #6642.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 1]
```diff
@@ -516,7 +516,8 @@ private data class ModuleInfoFile(
 }
 
 private fun ModuleInfo.toVcsInfo(): VcsInfo? {
-    val info = jsonMapper.readValue<ModuleInfoFile>(File(goMod).resolveSibling("$version.info"))
+    val infoFile = File(goMod).resolveSibling("$version.info")
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt]: Conventional commit nit: I guess the more correct type would be "refactor".

**Review by sschuberth — COMMENTED** [review]
I'll leave the approval to @oheger-bosch who proposed the alternative PR #6638.

File: analyzer/src/funTest/assets/projects/synthetic/gomod-unused-deps-expected-output.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+---
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/funTest/assets/projects/synthetic/gomod-unused-deps-expected-output.yml:1]: Making this unpersonal would IMO change the meaning. What do you suggest exactly?

**Review by oheger-bosch — DISMISSED** [review]
Changes look good to me. @MarcelBochtler, is there any chance you can test this against our problematic project? Otherwise, I am fine with merging this.

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 3]
```diff
@@ -484,7 +484,7 @@ private data class ModuleInfo(
     val main: Boolean = false,
 
     @JsonProperty("GoMod")
-    val goMod: String
+    val goMod: String?
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt:487]: Just checking: In @oheger-bosch's [proposal](https://github.com/oss-review-toolkit/ort/pull/6638/files#diff-d0e6f488a95a59913828fe32b76dae1b1ad197ddac08af8485ffa87578a0ae79R489) this also got a default value of `null`. Is this intentionally *not* done here so somehow express that `goMod` being `null` is rather an exceptional case than the default?

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/GoMod.kt:487]: This class is only construction in context of deserialization. There is no constructor calls in the code. So, a default null value is not needed.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/GoMod.kt:487]: Now I'm confused: Isn't the point that the `goMod` property might be missing from the JSON to deserialize, hence we'd need a default value here?

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/GoMod.kt:487]: Ah, seems like so. I've never made that really clear to me, it seems.

Given that in all other places where data class properties are nullable a default value of `null` *is* assigned, how about also aligning this as part of this PR? (In either way.)

**Review by sschuberth — APPROVED** [review]

**Review by MarcelBochtler — APPROVED** [review]

**Comment by fviernau** [comment]
Merged as the only failing test (ClearlyDefinedStorageFunTest) was unrelated.

**Pull request merged**